src/cli/cli.hpp

```diff
@@ -531,10 +533,31 @@ class Interpreter : public OutputImplementer, public Output
     static void HandleTimer(Timer &aTimer);
     void        HandleTimer(void);
 
-    const otCliCommand *mUserCommands;
-    uint8_t             mUserCommandsLength;
-    void               *mUserCommandsContext;
-    bool                mCommandIsPending;
+    class otCliCommandList : public LinkedListEntry<otCliCommandList>
+    {
+        friend class LinkedListEntry<otCliCommandList>;
+
+    public:
+        void Set(const otCliCommand *aCmdList, uint8_t aLength, void *aContext)
+        {
+            mCommands       = aCmdList;
+            mCommandsLength = aLength;
+            mContext        = aContext;
+        }
+
+        const otCliCommand *mCommands;
+        uint8_t             mCommandsLength;
+        void               *mContext;
+
+    private:
+        otCliCommandList *mNext;
+    };
+
+    otCliCommandList *AllocateCommandList(void) { return mCommandListPool.Allocate(); }
+
+    LinkedList<otCliCommandList>                   mUserCommands;
+    Pool<otCliCommandList, kMaxUserCommandEntries> mCommandListPool;
```

Comment:
I am not sure it is good idea to use `LinkedList` and `Pool` types here in `cli` module.

These types are intended for use by OT core modules and `cli` is app/next layer and not part of OT core. `cli` should only use the `ot` public definitions and APIs. I know there may be few  cases that this is not followed, but I feel we should use these types.